### ISSUE TITLE
Can't provide a default value in static method if parameter has a generic annotation

### ISSUE THREAD

**Issue by evtn** [microsoft/pyright#5571]
**Describe the bug**
Pyright complains about function param default value being a literal and thus (for some reason) unassignable, if parameter has a generic annotation

**Code or Screenshots**
If possible, provide a minimal, self-contained code sample (surrounded by triple back ticks) to demonstrate the issue. The code should define or import all referenced symbols.

```python
from __future__ import annotations
from typing import Generic, TypeVar


_T = TypeVar("_T", bound=int)


class Container(Generic[_T]):
    def __init__(self, x: _T):
        self.x = x

    @staticmethod
    def build(x: _T = 0) -> Container[_T]:
        return Container(x)
```
This shows:

```
13:23 - error: Expression of type "Literal[0]" cannot be assigned to parameter of type "_T@Container"
    Type "Literal[0]" cannot be assigned to type "_T@Container" (reportGeneralTypeIssues)
```

This only breaks on a static method, a function would pass type check:

```python
from __future__ import annotations
from typing import Generic, TypeVar


_T = TypeVar("_T", bound=int)


def build(x: _T = 0) -> Container[_T]:
    return Container(x)


class Container(Generic[_T]):
    def __init__(self, x: _T = 0):
        self.x = x
```

I expect `Container.build()` to return either `Container[Literal[0]]` or `Container[int]` (and I don't see how it is unreasonable)

**VS Code extension or command-line**
Sublime Text with lsp-pyright, also gives the same error in CLI

**Comment by erictraut** [comment]
The use of a default value for a parameter that is annotated with a class-scoped type variable is a tricky issue. Mypy doesn't allow it under any circumstance. Neither does TypeScript. There are good reasons to disallow this within a type system because it opens up a number of type safety holes.

Pyright didn't originally allow this either. It now allows it for constructors but still disallows it for non-constructor methods (regardless of whether they're static, class or instance methods).

When a constructor is called for a generic class, the class can be explicitly specialized, but it can also be implicitly specialized based on the arguments passed to the constructor. If it is explicitly specialized, then the provided arguments must be type consistent with the explicit specialization.

```python
v1 = list[int]()
reveal_type(v1) # list[int]

v2 = list([2, 3, 4])
reveal_type(v2) # list[int]

v3 = list[str]([2, 3, 4]) # Error
```

When you call a method that is not a constructor, the class or object is typically specialized prior to calling that method, and any type variables used within the implementation of that method must honor the specialized type. For that reason, pyright still disallows the use of default values when a class-scoped type variable is used in the parameter type.

Pyright allows default values to be used with function-scoped type variables in functions or methods. This is why your `build` function does not result in an error. Mypy does not allow default values in this case.

The error you're seeing here is consistent with the error you see with mypy for your top code sample. In your bottom code sample, mypy generates errors for both `build` and `Container.__init__`. Pyright does not generate errors in this case because it is demonstrably type safe to allow this.
